### PR TITLE
feat(router): Use `NTID` in `MIT` payments if the `pg_agnostic_mit` config is enabled

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

During the MIT process, there needs to be validation to check the off_session field. If set to true, then check if the pg_agnostic_mit config is enabled. If it is, perform a lookup in the payment methods table with the payment_method_id in the MIT payment request for the NRID.

If the confirm call fails, retry across the retry-able connectors specified in the `pg_agnostic_mit` configuration.

If the NRID is not present for that particular payment_method_id, or if the `pg_agnostic_mit` flag is not enabled, call the initial connector with which the CIT was performed.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Create a Merchant account
Create a mca for stripe, adyen or cybersource
Enable the pg_agnostic endpoint
```
curl --location 'https://sandbox.hyperswitch.io/routing/business_profile/:business_profile_id/configs/pg_agnostic_mit' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer JWT_Token' \
--data '{
    "enabled": true
}'
```
Create a mandate payment (Setup mandate)
image

```
{
    "amount": 0,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "singh",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "03",
            "card_exp_year": "2030",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        }
    },
    "setup_future_usage": "off_session",
    "payment_type": "setup_mandate",
        "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "127.0.0.1",
                "user_agent": "amet irure esse"
            }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "John",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "John",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "125.0.0.1"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {},
    "order_details": [
        {
            "product_name": "Apple iphone 15",
            "quantity": 1,
            "amount": 0,
            "account_name": "transaction_processing"
        }
    ]
}
```
DB screenshot where `network_transaction_id` is stored in the `payment_methods` table
image
Create a payment with confirm false
image
List payment methods for customer
```
curl --location 'http://localhost:8080/customers/:customer_id/payment_methods' \
--header 'Accept: application/json' \
--header 'api-key: api-key'
```
image
Confirm the payment by passing the connector as stripe in routing

```
{
    "amount": 300,
    "currency": "USD",
    "confirm": false,
    "customer_id": "cyb111",
    "return_url": "https://google.com",
    "email": "something@gmail.com",
    "name": "Joseph Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "statement_descriptor_name": "Juspay",
    "statement_descriptor_suffix": "Router",
    "off_session": true,
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "connector_metadata": {
        "noon": {
            "order_category": "pay"
        }
    },

    "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "125.0.0.1"
    }
}
```

```
curl --location 'http://localhost:8080/payments/{{payment_id}}/confirm' \
--header 'api-key: api-key' \
--header 'Content-Type: application/json' \
--data '{
    "payment_token": "payment_token",
    "client_secret": "client_secret",
    "payment_method": "card"
}'
```
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/stripe/transformers.rs [hunk 1]
```diff
@@ -1808,7 +1809,45 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for PaymentIntentRequest {
...
+                        | payments::PaymentMethodData::Crypto(_)
+                        | payments::PaymentMethodData::MandatePayment
+                        | payments::PaymentMethodData::Reward
+                        | payments::PaymentMethodData::Upi(_)
+                        | payments::PaymentMethodData::Voucher(_)
+                        | payments::PaymentMethodData::GiftCard(_)
+                        | api::PaymentMethodData::CardToken(_) => {
+                            Err(errors::ConnectorError::NotSupported {
+                                message: "Network tokenization for payment method".to_string(),
+                                connector: "Adyen",
```
Comment (Aprabhat19) [inline crates/router/src/connector/stripe/transformers.rs:1839]:
```suggestion
                                connector: "Stripe",
```

File: crates/router/src/core/payments.rs [hunk 2]
```diff
@@ -2991,42 +2996,120 @@ pub fn decide_multiplex_connector_for_normal_or_recurring_payment<F: Clone>(
...
+
+            let pg_agnostic = state
+                .store
+                .find_config_by_key_unwrap_or(
+                    &format!("pg_agnostic_mandate_{}", profile_id),
+                    Some("false".to_string()),
+                )
+                .await
+                .change_context(errors::ApiErrorResponse::ResourceIdNotFound)
+                .attach_printable("The conditional config was not found in the DB")?;
```
Comment (Aprabhat19) [inline crates/router/src/core/payments.rs:3036]: Throw a relevant error message

File: crates/router/src/core/payments.rs [hunk 3]
```diff
@@ -3079,6 +3152,28 @@ pub fn decide_multiplex_connector_for_normal_or_recurring_payment<F: Clone>(
     }
 }
 
+pub async fn is_network_transaction_id_flow(
+    state: &AppState,
+    pg_agnostic: &String,
+    connector: enums::Connector,
+    payment_method_info: &storage::PaymentMethod,
+) -> RouterResult<bool> {
```
Comment (Aprabhat19) [inline crates/router/src/core/payments.rs:3160]: Can we just return bool here. Also this  fn need not be async.

File: crates/router/src/core/payments.rs [hunk 4]
```diff
@@ -3079,6 +3152,28 @@ pub fn decide_multiplex_connector_for_normal_or_recurring_payment<F: Clone>(
...
+
+    if pg_agnostic == "true"
+        && payment_method_info.payment_method == storage_enums::PaymentMethod::Card
+        && ntid_supported_connectors.contains(&connector)
+        && payment_method_info.network_transaction_id.is_some()
+    {
+        Ok(true)
+    } else {
+        Ok(false)
+    }
```
Comment (Aprabhat19) [inline crates/router/src/core/payments.rs:3174]:
```suggestion
   return pg_agnostic == "true"
        && payment_method_info.payment_method == storage_enums::PaymentMethod::Card
        && ntid_supported_connectors.contains(&connector)
        && payment_method_info.network_transaction_id.is_some()
   
```

**Review by Aprabhat19 — CHANGES_REQUESTED** [review]

**Review by Aprabhat19 — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

File: crates/router/src/core/payments/helpers.rs [hunk 5]
```diff
@@ -3419,10 +3419,12 @@ impl AttemptType {
 
 pub async fn update_payment_method_with_ntid(
     state: &AppState,
+    pg_agnostic_config: &String,
     pm: Option<storage::PaymentMethod>,
     payment_response: Result<PaymentsResponseData, ErrorResponse>,
 ) -> CustomResult<(), errors::ApiErrorResponse> {
-    let network_transaction_id = payment_response
+    if pg_agnostic_config == "true" {
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/helpers.rs:3426]: can this be boolean?

File: crates/router/src/connector/stripe/transformers.rs [hunk 6]
```diff
@@ -3187,6 +3231,7 @@ pub struct Card {
 // pub struct Card
 #[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Default, Eq, PartialEq)]
 pub struct StripeMandateOptions {
+    #[serde(rename = "payment_method_options[card][mandate_options]")]
```
[the remark this answers is not in the document]
Comment (ShankarSinghC) [reply crates/router/src/connector/stripe/transformers.rs:3234]: No, this is not being used in the existing flows.

File: crates/router/src/core/payments/helpers.rs (hunk 5 above)
Comment (ShankarSinghC) [reply crates/router/src/core/payments/helpers.rs:3426]: Can be taken in the next pr.

**Review by jagan-jaya — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

**Pull request merged**